base-files: sysupgrade: deal with spaces in path

luizluca opened the pull request:
It was part of #2552, which is stalled for too many time. I'll try one at a time.

aparcar commented:
Hi, thanks for the patch. Could you please separate both fixes into single commits for easier review?

luizluca committed:
base-files: sysupgrade: fix backup of files with spaces in path

Fix backup of files with spaces in path.

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca committed:
base-files: sysupgrade: fix backup file (-f) with space in path

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

luizluca commented on package/base-files/files/sbin/sysupgrade:
```diff
@@ -118,32 +118,41 @@ list_conffiles() {
 		!/^ / { conffiles = 0; next }
-		conffiles == 1 { print }
+		# in format: <space><filename><space><hash>
+		conffiles == 1 { print $(NF) "  " substr($0,2,length($0)-length($(NF))-2) }
```
list_conffiles here will return the same format sha256sum returns (and expects while checking a hash)

luizluca replied:
list_conffiles() will return the same format sha256sum returns (and expects while checking a hash).
This avoids problems with spaces and rebuilding the expected sha256sum aferwards in list_changed_conffiles().

luizluca commented on package/base-files/files/sbin/sysupgrade:
```diff
@@ -118,32 +118,41 @@ list_conffiles() {
-	local filter=$1
+find_bkpfiles() {
+	find "$@" \( -type f -o -type l \) \
+		${SKIP_UNCHANGED:+\( \! -exec test -e "/rom/{}" \; -o \! -exec cmp -s "/{}" "/rom/{}" \; \)} \
```
SKIP_UNCHANGED was used to define find_filter. However, it require to expand a string with spaces without quotes. It is not really affected by "spaces in config files" but a ${SKIP_UNCHANGED:+} expansion is safer.

luizluca commented on package/base-files/files/sbin/sysupgrade:
```diff
@@ -118,32 +118,41 @@ list_conffiles() {
+		while read -r filename; do
+			set -- "$@" "$filename"
+		done
+		find_bkpfiles "$@" 2>/dev/null
```
find_bkpfiles is equivalent to the old find $(sed...) but it receives the filenames from args and $filter is included inside find_bkpfiles reading directly SKIP_UNCHANGED

luizluca commented on package/base-files/files/sbin/sysupgrade:
```diff
@@ -157,15 +166,15 @@ add_overlayfiles() {
 		local keepfiles=$1.keepfiles
 
-		list_conffiles | cut -f2 -d ' ' | sort -u > "$conffiles"
+		list_conffiles | cut -f3- -d ' ' | sort -u > "$conffiles"
```
Just adapts to the new list_conffiles format (<hash><space><space><filename with spaces>)

luizluca commented on package/base-files/files/sbin/sysupgrade:
```diff
@@ -14,7 +14,7 @@ export SAVE_OVERLAY=0
 export SAVE_PARTITIONS=1
 export SAVE_INSTALLED_PKGS=0
-export SKIP_UNCHANGED=0
+export SKIP_UNCHANGED=
```
SKIP_UNCHANGED needs to be unset to add args with ${SKIP_UNCHANGED:+arg1 arg2...}

luizluca commented on package/base-files/files/sbin/sysupgrade:
```diff
@@ -175,15 +184,16 @@ add_overlayfiles() {
-		} |  grep -v -x -F -f $conffiles |
-		     grep -v -x -F -f $keepfiles | sort -u > "$packagesfiles"
+		} |  grep -v -x -F -f "$conffiles" |
+		     grep -v -x -F -f "$keepfiles" | sort -u > "$packagesfiles"
```
These series of quotes are not really necessary but they silence shellcheck.

luizluca commented:
@aparcar , thanks for the interest in my PR.

I rebased it to fix the conflicts and split the two "fixes" in two commits. However, the second fix is trivial. I could split list_changed_conffiles/list_conffiles into a new commit if needed. There are some extra quotes added that are not really necessary and I could remove them if needed. The rest is a change on listing and processing paths using shell args instead of unquoted string expansion. 93b612221208c731da9b6c7100af636894405f75 (not mine) already made it smaller.

I added some comments to guide the review. I can squash back all the commits before merge if needed.

adschm commented:
Looks complicated and error-prone to me. Is it really worth it, compared to a "no-spaces-in-file-names" policy?

luizluca commented:
> Looks complicated and error-prone to me. Is it really worth it, compared to a "no-spaces-in-these-file-names" policy?

It's a bit complicate to create shell scripts that correctly deal with spaces and special symbols, specially with ash.
The safest way to solve it is to use null terminated strings (find -print0, xargs -0) and args (as ash does not have arrays), and the code will not be pretty. However, it's the only way to be really safe. A "new line", for example, is a perfectly acceptable char for a filename. This PR simply ignored that case because there are so many things that will break with a new line, even before sysupgrade backups. Is there any specific change that can and needs to be simplified?

I do not agree with the "error-prone" argument. It's strange to keep a code that has a known bug instead of a new one that might have an error. I'm proposing this PR because I lost files (with spaces) that I expected to be in a backup.

From someone that do shell injections "just for fun", "no-spaces-in-these-file-names" policy is just exactly what I target. It normally indicates a strings expansion without quotes, which is even worse than quoted without escaping. If there is a "no space" policy, it should be enforced by code and not assumed to be respected. However, I think that failing a backup because it contains a filename with space is just a lazy solution. When you backup a directory that a non-admin might affect the filename (like a samba share), the user will not give a damn about OpenWrt policies.

ehem reviewed (approved):
Being prepared for someone choosing odd filenames is always a Good Thing.
